pkg/envoy/xds/server_e2e_test.go

```diff
@@ -0,0 +1,419 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xds
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/cilium/cilium/pkg/envoy/api"
+
+	"github.com/golang/protobuf/proto"
+	. "gopkg.in/check.v1"
+)
+
+// Hook up gocheck into the "go test" runner.
+func Test(t *testing.T) {
+	//logging.ToggleDebugLogs(true)
+	TestingT(t)
+}
+
+type ServerSuite struct{}
+
+var _ = Suite(&ServerSuite{})
+
+const (
+	TestTimeout      = 10 * time.Second
+	StreamTimeout    = 2 * time.Second
+	CacheUpdateDelay = 250 * time.Millisecond
+)
+
+var resources = []*api.RouteConfiguration{
+	&api.RouteConfiguration{
+		Name: "resource0",
+	},
+	&api.RouteConfiguration{
+		Name: "resource1",
+	},
+	&api.RouteConfiguration{
+		Name: "resource2",
+	},
+}
+
+func (s *ServerSuite) TestRequestAllResources(c *C) {
+	typeURL := "type.googleapis.com/envoy.api.v2.DummyConfiguration"
+
+	var err error
+	var req *api.DiscoveryRequest
+	var resp *api.DiscoveryResponse
+
+	ctx, cancel := context.WithTimeout(context.Background(), TestTimeout)
+	defer cancel()
+
+	cache := NewCache()
+
+	streamCtx, closeStream := context.WithCancel(ctx)
+	stream := NewMockStream(streamCtx, 1, 1, StreamTimeout, StreamTimeout)
+	defer stream.Close()
+
+	server := NewServer(map[string]ObservableResourceSet{
+		typeURL: cache,
+	}, TestTimeout)
+
+	streamDone := make(chan struct{})
+
+	// Run the server's stream handler concurrently.
+	go func() {
+		err := server.HandleRequestStream(ctx, stream, AnyTypeURL)
+		close(streamDone)
+		c.Check(err, IsNil)
+	}()
+
+	// Request all resources.
+	req = &api.DiscoveryRequest{
+		TypeUrl:       typeURL,
+		VersionInfo:   "",
+		Node:          nil, // Ignored by server.
+		ResourceNames: nil,
+		ResponseNonce: "",
+	}
+	err = stream.SendRequest(req)
+	c.Assert(err, IsNil)
+
+	// Expecting an empty response.
```

Comment:
Can we write table-driven tests here?

I see that the majority of that file repeats the following block:

```
resp, err = stream.RecvResponse()
c.Assert(err, IsNil)
c.Assert(resp.VersionInfo, Equals, ?)
c.Check(resp.Resources, HasLen, ?)
c.Check(resp.Canary, Equals, false)
c.Check(resp.TypeUrl, Equals, typeURL)
c.Assert(resp.Nonce, Not(Equals), "")
```

where usually `VersionInfo` and `Resources` differ. I think we might write those tests in the following way:

```
tests := []struct {
	versionInfo string
	resourcesLen int
}{
	{"0", 0},
	{"1", 1},
	{"2", 2}
}

for _, tc .= range tests {
	req.VersionInfo = ...
	req.ResponseNonce = resp.Nonce
	err = stream.SendRequest(req)

 	c.Assert(err, IsNil)
	c.Assert(resp.VersionInfo, Equals, tc.versionInfo)
	c.Assert(resp.Resources, HasLEn, tc.resourcesLen)
}
```

or something like that. Is it ok for you? Sorry if I'm missing some point which would make table-driven tests impossible.

Reply:
That's a priori a good idea, but it seems hard to generalize in this case. The xDS protocol is not an RPC, i.e. there is not always a response sent for every request. The tests here demonstrate and test that behavior.

One thing I'll look into is writing a Checker that matches on a response. That would make the code easier to read.

Reply:
I've implemented a `ResponseMatches` checker, which makes the tests more readable and concise, e.g.:
```
c.Assert(resp, ResponseMatches, "2", []proto.Message{resources[1]}, false, typeURL)
```